site/app/controllers/HomePageController.php

```diff
@@ -40,6 +42,79 @@ public function showLibrary() {
 		);
     }
 
+    /**
+     * @Route("/homework/library/search", methods={"GET"})
+     * @return Response
+     */
+    public function searchLibrary($path = null) {
+        if ($path === null) {
+            $path = $this->homework_path;
+        }
+        $gradeable_ids = FileUtils::getDirWithText($path, "config.json");
+        return Response::JsonOnlyResponse(
+            JsonResponse::getSuccessResponse($gradeable_ids)
+        );
+	}
+
+    /**
+     * @Route("/homework/library/search/{query}", methods={"GET"})
+     * @return Response
+     */
+    public function searchLibraryWithQuery($query, $path = null) {
+        if ($path === null) {
+            $path = $this->homework_path;
+        }
+        $gradeable_ids = FileUtils::getPathsWithQuery($path, $query);
+        return Response::JsonOnlyResponse(
+            JsonResponse::getSuccessResponse($gradeable_ids)
+        );
+	}
+
+    /**
+     * @Route("/homework/library/search_gradeable/{query}", methods={"GET"})
+     * @return Response
+     */
+    public function searchLibraryGradeableWithQuery($query, $path = null) {
+        if ($path === null) {
+            $path = $this->homework_path;
+        }
+        $config = $this->getDetails($path, $query);
+        return Response::JsonOnlyResponse(
+            JsonResponse::getSuccessResponse($config)
+        );
+	}
+
+    public static function getDetails($path, $query) {
+        $config_path = FileUtils::getPathWithQueryAndTip($path, $query, 'config.json');
+        $readme_path = FileUtils::getPathWithQueryAndTip($path, $query, 'README.md');
```

Comment:
Are you ever grabbing the readme file in another request? If not, do you even need to determine if it exists and get the path?

Reply:
The default value if a README does not exist is false